desktop/cmp/treemap/TreeMapModel.js

```diff
@@ -0,0 +1,405 @@
+     */
+    constructor({
+        store,
+        gridModel,
+        maxNodes = 1000,
+        highchartsConfig,
+        labelField = 'name',
+        valueField = 'value',
+        heatField = 'value',
+        maxDepth,
+        algorithm = 'squarified',
+        colorMode = 'linear',
+        onClick,
+        onDoubleClick,
+        tooltip = true,
+        emptyText,
+        filter
+    } = {}) {
+        this.gridModel = gridModel;
+        this.store = store ? store : gridModel ? gridModel.store : null;
+        throwIf(!this.store,  'TreeMapModel requires either a Store or a GridModel');
+        this.maxNodes = maxNodes;
+
+        this.highchartsConfig = highchartsConfig;
+        this.labelField = labelField;
+        this.valueField = valueField;
+        this.heatField = heatField;
+        this.maxDepth = maxDepth;
+
+        throwIf(!['sliceAndDice', 'stripes', 'squarified', 'strip'].includes(algorithm), `Algorithm "${algorithm}" not recognised.`);
+        this.algorithm = algorithm;
+
+        throwIf(!['linear', 'balanced', 'none'].includes(colorMode), `Color mode "${colorMode}" not recognised.`);
+        this.colorMode = colorMode;
+
+        this.onClick = withDefault(onClick, this.defaultOnClick);
+        this.onDoubleClick = withDefault(onDoubleClick, this.defaultOnDoubleClick);
+        this.tooltip = tooltip;
+        this.emptyText = emptyText;
+
+        this._filter = filter;
+
+        this.addReaction({
+            track: () => [
+                this.store.rootRecords,
+                this.expandState,
+                this.colorMode,
+                this.labelField,
+                this.valueField,
+                this.heatField,
+                this.maxDepth
+            ],
+            run: ([rawData]) => this.data = this.processData(rawData)
+        });
+    }
+
+    @computed
+    get total() {
+        return sumBy(this.data, it => {
+            // Only include root records that pass the filter
+            if (it.parent || (this._filter && !this._filter(it.record))) return 0;
+            return it.value;
+        });
+    }
+
+    @computed
+    get valueFieldLabel() {
+        const field = this.store.fields.find(it => it.name === this.valueField);
+        return field ? field.label : this.valueField;
+    }
+
+    @computed
+    get heatFieldLabel() {
+        const field = this.store.fields.find(it => it.name === this.heatField);
+        return field ? field.label : this.heatField;
+    }
+
+    @computed
+    get selectedIds() {
+        if (!this.gridModel || this.gridModel.selModel.mode === 'disabled') return [];
+        return this.gridModel.selModel.ids;
+    }
+
+    @computed
+    get expandState() {
+        if (!this.gridModel || !this.gridModel.treeMode) return {};
+        return this.gridModel.expandState;
+    }
+
+    @computed
+    get hasData() {
+        return !isEmpty(this.data);
+    }
+
+    @computed
```

Comment:
This seems like an interesting way to recognize an 'error'.  Wondering about the name of this? Would like to discuss at some point

Reply:
I had the same thought - definitely an interesting pattern for these components we like to write where they have an error panel that will get rendered if the entire thing is generally not in a working state for whatever reason.  (Something I feel like we could build into more components in a standard way...)

Thought had crossed my mind that `errors` or `errorMessages` might be a bit of a better name and have it always return an array. We could even author a small `ErrorMessages` component that took that array and rendered a list with a standard CSS class...  (Could imagine that getting fancier in some future iteration to truncate messages, optionally disclose exception stacks, etc)